clusters/resource_cluster.go

```diff
@@ -58,8 +59,72 @@ func ZoneDiffSuppress(k, old, new string, d *schema.ResourceData) bool {
 	return false
 }
 
+// This method is a duplicate of Validate() in clusters/clusters_api.go that uses Go SDK.
+// Long term, Validate() in clusters_api.go will be removed once all the resources using clusters are migrated to Go SDK.
+func Validate(cluster compute.CreateCluster) error {
+	if cluster.NumWorkers > 0 || cluster.Autoscale != nil {
+		return nil
+	}
+	profile := cluster.SparkConf["spark.databricks.cluster.profile"]
+	master := cluster.SparkConf["spark.master"]
+	resourceClass := cluster.CustomTags["ResourceClass"]
+	if profile == "singleNode" && strings.HasPrefix(master, "local") && resourceClass == "SingleNode" {
+		return nil
+	}
+	return fmt.Errorf("NumWorkers could be 0 only for SingleNode clusters. See https://docs.databricks.com/clusters/single-node.html for more details")
+}
+
+// This method is a duplicate of ModifyRequestOnInstancePool() in clusters/clusters_api.go that uses Go SDK.
+// Long term, ModifyRequestOnInstancePool() in clusters_api.go will be removed once all the resources using clusters are migrated to Go SDK.
+func ModifyRequestOnInstancePool(cluster *compute.CreateCluster) {
+	// Instance profile id does not exist or not set
+	if cluster.InstancePoolId == "" {
+		// Worker must use an instance pool if driver uses an instance pool,
+		// therefore empty the computed value for driver instance pool.
+		cluster.DriverInstancePoolId = ""
+		return
+	}
+	if cluster.AwsAttributes != nil {
+		// Reset AwsAttributes
+		awsAttributes := compute.AwsAttributes{
+			InstanceProfileArn: cluster.AwsAttributes.InstanceProfileArn,
+		}
+		cluster.AwsAttributes = &awsAttributes
+	}
+	if cluster.AzureAttributes != nil {
+		cluster.AzureAttributes = &compute.AzureAttributes{}
+	}
+	if cluster.GcpAttributes != nil {
+		gcpAttributes := compute.GcpAttributes{
+			GoogleServiceAccount: cluster.GcpAttributes.GoogleServiceAccount,
+		}
+		cluster.GcpAttributes = &gcpAttributes
+	}
+	cluster.EnableElasticDisk = false
+	cluster.NodeTypeId = ""
+	cluster.DriverNodeTypeId = ""
+}
+
+// This method is a duplicate of FixInstancePoolChangeIfAny(d *schema.ResourceData) in clusters/clusters_api.go that uses Go SDK.
+// Long term, FixInstancePoolChangeIfAny(d *schema.ResourceData) in clusters_api.go will be removed once all the resources using clusters are migrated to Go SDK.
+// https://github.com/databricks/terraform-provider-databricks/issues/824
+func FixInstancePoolChangeIfAny(d *schema.ResourceData, cluster compute.CreateCluster) {
+	oldInstancePool, newInstancePool := d.GetChange("instance_pool_id")
+	oldDriverPool, newDriverPool := d.GetChange("driver_instance_pool_id")
+	if oldInstancePool != newInstancePool &&
+		oldDriverPool == oldInstancePool &&
+		oldDriverPool == newDriverPool {
+		cluster.DriverInstancePoolId = cluster.InstancePoolId
+	}
+}
+
+type LibraryList struct {
```

Comment:
We need this type (with struct annotations) below, so we have to explicitly declare it. 
`````
var libraryList LibraryList
common.DataToStructPointer(d, clusterSchema, &libraryList)
`````
Given it's only used in clusters and to not encourage wide use, I think we should keep it here as compared to libraries_api.

Reply:
cc: @miles, I am not sure if there is a better / easier way to do this, we could probably get the type from ClusterSpec.Libraries using reflection but I think having a 2 line struct would be simpler.

Reply:
Updated to use `ClusterSpec`